File: Code_Examples/Chapter_04/laplace_openmp.f90
program serial
      implicit none

      integer, parameter             :: width=1000
      integer, parameter             :: height=1000
      double precision, parameter    :: temp_tolerance=0.01

      integer                        :: i, j, iteration=1
      double precision               :: worst_dt=100.0
      real                           :: start_time, stop_time

      double precision, dimension(0:height+1,0:width+1) :: temperature, temperature_previous

      call cpu_time(start_time)

      call initialize(temperature_previous)

      do while ( worst_dt > temp_tolerance )

         !$omp parallel do
         do j=1,width
            do i=1,height
               temperature(i,j)=0.25*(temperature_previous(i+1,j)+temperature_previous(i-1,j)+ &
                                      temperature_previous(i,j+1)+temperature_previous(i,j-1) )
            enddo
         enddo
         !$omp end parallel do

         worst_dt=0.0

         !$omp parallel do reduction(max:worst_dt)
         do j=1,width
            do i=1,height
               worst_dt = max( abs(temperature(i,j) - temperature_previous(i,j)), worst_dt )
               temperature_previous(i,j) = temperature(i,j)
            enddo
         enddo
         !$omp end parallel do

         if( mod(iteration,100).eq.0 ) then
            call track_progress(temperature, iteration)
         endif

         iteration = iteration+1

      enddo

      call cpu_time(stop_time)

      print*, 'Max error at iteration ', iteration-1, ' was ',worst_dt
      print*, 'Total time was ',stop_time-start_time, ' seconds.'

end program serial


subroutine initialize( temperature_previous )
      implicit none

      integer, parameter             :: width=1000
      integer, parameter             :: height=1000
      integer                        :: i,j

      double precision, dimension(0:height+1,0:width+1) :: temperature_previous

      temperature_previous = 0.0

      do i=0,height+1
         temperature_previous(i,0) = 0.0
         temperature_previous(i,width+1) = (100.0/height) * i
      enddo

      do j=0,width+1
         temperature_previous(0,j) = 0.0
         temperature_previous(height+1,j) = ((100.0)/width) * j
      enddo

end subroutine initialize


subroutine track_progress(temperature, iteration)
      implicit none

      integer, parameter             :: width=1000
      integer, parameter             :: height=1000
      integer                        :: i,iteration

      double precision, dimension(0:height+1,0:width+1) :: temperature

      print *, '---------- Iteration number: ', iteration, ' ---------------'
      do i=5,0,-1
         write (*,'("("i4,",",i4,"):",f6.2,"  ")',advance='no') &
                   height-i,width-i,temperature(height-i,width-i)
      enddo
      print *
end subroutine track_progress

File: Code_Examples/Chapter_04/laplace_openmp.c
#include <stdlib.h>
#include <stdio.h>
#include <math.h>
#include <sys/time.h>

#define WIDTH      1000
#define HEIGHT     1000

#define TEMP_TOLERANCE 0.01

double Temperature[HEIGHT+2][WIDTH+2];
double Temperature_previous[HEIGHT+2][WIDTH+2];

void initialize();
void track_progress(int iter);


int main(int argc, char *argv[]) {

    int i, j;
    int iteration=1;
    double worst_dt=100;
    struct timeval start_time, stop_time, elapsed_time;

    gettimeofday(&start_time,NULL);

    initialize();

    while ( worst_dt > TEMP_TOLERANCE ) {

        #pragma omp parallel for private(i,j)
        for(i = 1; i <= HEIGHT; i++) {
            for(j = 1; j <= WIDTH; j++) {
                Temperature[i][j] = 0.25 * (Temperature_previous[i+1][j] + Temperature_previous[i-1][j] +
                                            Temperature_previous[i][j+1] + Temperature_previous[i][j-1]);
            }
        }
        
        worst_dt = 0.0;

        #pragma omp parallel for reduction(max:worst_dt) private(i,j)
        for(i = 1; i <= HEIGHT; i++){
            for(j = 1; j <= WIDTH; j++){
	      worst_dt = fmax( fabs(Temperature[i][j]-Temperature_previous[i][j]), worst_dt);
	      Temperature_previous[i][j] = Temperature[i][j];
            }
        }

        if((iteration % 100) == 0) {
 	    track_progress(iteration);
        }

	iteration++;
    }

    gettimeofday(&stop_time,NULL);
    timersub(&stop_time, &start_time, &elapsed_time);

    printf("\nMax error at iteration %d was %f\n", iteration-1, worst_dt);
    printf("Total time was %f seconds.\n", elapsed_time.tv_sec+elapsed_time.tv_usec/1000000.0);

}


void initialize(){

    int i,j;

    for(i = 0; i <= HEIGHT+1; i++){
        for (j = 0; j <= WIDTH+1; j++){
            Temperature_previous[i][j] = 0.0;
        }
    }

    for(i = 0; i <= HEIGHT+1; i++) {
        Temperature_previous[i][0] = 0.0;
        Temperature_previous[i][WIDTH+1] = (100.0/HEIGHT)*i;
    }
    
    for(j = 0; j <= WIDTH+1; j++) {
        Temperature_previous[0][j] = 0.0;
        Temperature_previous[HEIGHT+1][j] = (100.0/WIDTH)*j;
    }
}


void track_progress(int iteration) {

    int i;

    printf("---------- Iteration number: %d ------------\n", iteration);
    for(i = HEIGHT-5; i <= HEIGHT; i++) {
        printf("[%d,%d]: %5.2f  ", i, i, Temperature[i][i]);
    }
    printf("\n");
}
